primitives/ethereum-poa/src/exchange.rs

```diff
@@ -0,0 +1,125 @@
+
+use codec::{Decode, Encode, EncodeLike};
+use frame_support::RuntimeDebug;
+use sp_std::marker::PhantomData;
+
+/// All errors that may happen during exchange.
+#[derive(RuntimeDebug, PartialEq)]
+pub enum Error {
+	/// Invalid peer blockchain transaction provided.
+	InvalidTransaction,
+	/// Peer transaction has invalid amount.
+	InvalidAmount,
+	/// Peer transaction has invalid recipient.
+	InvalidRecipient,
+	/// Cannot map from peer recipient to this blockchain recipient.
+	FailedToMapRecipients,
+	/// Failed to convert from peer blockchain currency to this blockhain currency.
+	FailedToConvertCurrency,
+	/// Deposit has failed.
+	DepositFailed,
+}
+
+/// Result of all exchange operations.
+pub type Result<T> = sp_std::result::Result<T, Error>;
+
+/// Peer blockchain lock funds transaction.
+#[derive(Encode, Decode, Clone, RuntimeDebug, PartialEq, Eq)]
+pub struct LockFundsTransaction<TransferId, Recipient, Amount> {
+	/// Something that uniquely identifies this transfer.
+	pub id: TransferId,
+	/// Funds recipient on the peer chain.
+	pub recipient: Recipient,
+	/// Amount of the locked funds.
+	pub amount: Amount,
+}
+
+/// Peer blockchain transaction that may represent lock funds transaction.
+pub trait MaybeLockFundsTransaction {
+	/// Transaction type.
+	type Transaction;
+	/// Identifier that uniquely identifies this transfer.
+	type Id: Decode + Encode + EncodeLike;
+	/// Peer recipient type.
+	type Recipient;
+	/// Peer currency amount type.
+	type Amount;
+
+	/// Parse lock funds transaction of the peer blockchain. Returns None if
+	/// transaction format is unknown, or it isn't a lock funds transaction.
+	fn parse(tx: &Self::Transaction) -> Result<LockFundsTransaction<Self::Id, Self::Recipient, Self::Amount>>;
+}
+
+/// Map that maps recipients from peer blockchain to this blockchain recipients.
+pub trait RecipientsMap {
+	/// Peer blockchain recipient type.
+	type PeerRecipient;
+	/// Current blockchain recipient type.
+	type Recipient;
+
+	/// Lookup current blockchain recipient by peer blockchain recipient.
+	fn map(peer_recipient: Self::PeerRecipient) -> Result<Self::Recipient>;
+}
+
+/// Conversion between two currencies.
+pub trait CurrencyConverter {
+	/// Type of the source currency amount.
+	type SourceAmount;
+	/// Type of the target currency amount.
+	type TargetAmount;
+
+	/// Covert from source to target currency.
+	fn convert(amount: Self::SourceAmount) -> Result<Self::TargetAmount>;
+}
+
+/// Currency deposit.
+pub trait DepositInto {
+	/// Recipient type.
+	type Recipient;
+	/// Currency amount type.
+	type Amount;
+
+	/// Grant some money to given account.
+	fn deposit_into(recipient: Self::Recipient, amount: Self::Amount) -> Result<()>;
+}
+
+/// Recipients map which is used when accounts ids are the same on both chains.
+#[derive(Debug)]
+pub struct IdentityRecipients<AccountId>(PhantomData<AccountId>);
+
+impl<AccountId> RecipientsMap for IdentityRecipients<AccountId> {
+	type PeerRecipient = AccountId;
```

Comment:
I'm a bit confused with the naming here. Is a `PeerRecipient` the same as a `Sender` from another chain?

Reply:
It is decided by the combination of `MaybeLockFundsTransaction` and `RecipientsMap`  implementations.